src/NCalc/Visitors/LambdaExpressionVisitor.cs

```diff
@@ -0,0 +1,460 @@
+        Type paramsElementType = null;
+        var paramsParameterPosition = 0;
+        if (!hasParamsKeyword)
+        {
+            if (parameters.Length != arguments.Length) return null;
+        }
+        else
+        {
+            paramsParameterPosition = lastParameter.Position;
+            paramsElementType = lastParameter.ParameterType.GetElementType();
+            paramsKeywordArgument = new LinqExpression[arguments.Length - parameters.Length + 1];
+        }
+
+        var functionMemberScore = 0;
+        for (var i = 0; i < arguments.Length; i++)
+        {
+            var isParamsElement = hasParamsKeyword && i >= paramsParameterPosition;
+            var argument = arguments[i];
+            var argumentType = argument.Type;
+            var parameterType = isParamsElement ? paramsElementType : parameters[i].ParameterType;
+            if (argumentType != parameterType)
+            {
+                var canCastImplicitly = TryCastImplicitly(argumentType, parameterType, ref argument);
+                if (!canCastImplicitly) return null;
+                functionMemberScore++;
+            }
+            if (!isParamsElement)
+            {
+                newArguments[i] = argument;
+            }
+            else
+            {
+                paramsKeywordArgument[i - paramsParameterPosition] = argument;
+            }
+        }
+
+        if (hasParamsKeyword)
+        {
+            newArguments[paramsParameterPosition] = LinqExpression.NewArrayInit(paramsElementType, paramsKeywordArgument);
+        }
+        return Tuple.Create(functionMemberScore, newArguments);
+    }
+
+    private bool TryCastImplicitly(Type from, Type to, ref LinqExpression argument)
+    {
+        var convertingFromPrimitiveType = TypeHelper.ImplicitPrimitiveConversionTable.TryGetValue(from, out var possibleConversions);
+        if (!convertingFromPrimitiveType || !possibleConversions.Contains(to))
+        {
+            argument = null;
+            return false;
+        }
+        argument = LinqExpression.Convert(argument, to);
+        return true;
+    }
+
+    private LinqExpression WithCommonNumericType(LinqExpression left, LinqExpression right,
+        Func<LinqExpression, LinqExpression, LinqExpression> action, BinaryExpressionType expressiontype = BinaryExpressionType.Unknown)
+    {
+        left = UnwrapNullable(left);
+        right = UnwrapNullable(right);
+
+        if (_options.HasOption(ExpressionOptions.AllowBooleanCalculation))
+        {
+            if (left.Type == typeof(bool))
+            {
+                left = LinqExpression.Condition(left, LinqExpression.Constant(1.0), LinqExpression.Constant(0.0));
+            }
+
+            if (right.Type == typeof(bool))
+            {
+                right = LinqExpression.Condition(right, LinqExpression.Constant(1.0), LinqExpression.Constant(0.0));
+            }
+        }
+
+        var precedence = new[]
```

Comment:
If this variable never changes, we have move it to a `HashSet<string>` at `TypeHelper` right?

Reply:
As we need an order the HashSet is not the best way as the index in HashSet is meaningless

Reply:
@Bykiev agree, based on NCalc2 I changed the arrays of TypeHelper to HashSet, can you revert please?

Reply:
Added the changes in https://github.com/ncalc/ncalc/pull/188/commits/d29d74bd7b9cb14621339f809b42cca81f5f41f8

Reply:
I don't like the logic of `GetMostPreciseNumberType`, should we make it the same as `GetMostPreciseType`?

Reply:
I am on the phone and can't see much, but I agree in the lambda evaluation to re-use the most possible of our original visitor

Reply:
The logic is quite different, I can't say for now if such change is ok. Let's leave it as is.